source/io/Stream.cpp

```diff
@@ -13,132 +13,172 @@
-                                                                       s_aws_input_stream_cpp_destroy};
+                m_good = true;
+                m_underlying_stream->impl = this;
+                m_underlying_stream->allocator = m_allocator;
+                m_underlying_stream->vtable = &s_vtable;
+            }
 
-namespace Aws
-{
-    namespace Crt
-    {
-        namespace Io
-        {
-            struct aws_input_stream *AwsInputStreamNewCpp(
-                const std::shared_ptr<Aws::Crt::Io::IStream> &stream,
+            StdIOStreamInputStream::StdIOStreamInputStream(
+                std::shared_ptr<Aws::Crt::Io::IStream> stream,
                 Aws::Crt::Allocator *allocator) noexcept
+                : InputStream(allocator), m_stream(std::move(stream))
             {
-                struct aws_input_stream *input_stream = NULL;
-                struct aws_input_stream_cpp_impl *impl = NULL;
-
-                aws_mem_acquire_many(
-                    allocator,
-                    2,
-                    &input_stream,
-                    sizeof(struct aws_input_stream),
-                    &impl,
-                    sizeof(struct aws_input_stream_cpp_impl));
-
-                if (!input_stream)
+            }
+
+            bool StdIOStreamInputStream::Read(ByteBuf &buffer) noexcept
+            {
+                m_stream->read(reinterpret_cast<char *>(buffer.buffer + buffer.len), buffer.capacity - buffer.len);
+                buffer.len += static_cast<size_t>(m_stream->gcount());
+
+                return true;
+            }
+
+            StreamStatus StdIOStreamInputStream::GetStatus() const noexcept
+            {
+                StreamStatus status;
+                status.is_end_of_stream = m_stream->eof();
+                status.is_valid = m_stream->good();
+
+                return status;
+            }
+
+            int64_t StdIOStreamInputStream::GetLength() const noexcept
+            {
+                auto currentPosition = m_stream->tellg();
+
+                m_stream->seekg(0, std::ios_base::end);
+                int64_t retVal = -1;
+
+                if (m_stream->good())
                 {
-                    return NULL;
+                    retVal = static_cast<int64_t>(m_stream->tellg());
                 }
 
-                AWS_ZERO_STRUCT(*input_stream);
-                AWS_ZERO_STRUCT(*impl);
+                m_stream->seekg(currentPosition);
+
+                return retVal;
+            }
 
-                input_stream->allocator = allocator;
-                input_stream->vtable = &s_aws_input_stream_cpp_vtable;
-                input_stream->impl = impl;
+            bool StdIOStreamInputStream::Seek(OffsetType offsetType, SeekBasis seekBasis) noexcept
+            {
+                // very important, otherwise the stream can't be reused after reading the entire stream the first time.
+                m_stream->clear();
```

Comment:
Don't think this is necessary?
From [seekg docs](https://en.cppreference.com/w/cpp/io/basic_istream/seekg)
> Before doing anything else, seekg clears eofbit. | (since C++11)

Reply:
It’s necessary, this is a day long but tracking result.

Reply:
So is the quote I pulled from cppreference not correct? Or is not correctly implemented on some specific platform? I'm just curious

Reply:
I just know that when it wasn't there, seek didn't work ever again. And it doesn't hurt anything to do it.